src/service/xhr-impl.js

```diff
@@ -197,6 +219,40 @@ export class Xhr {
       assertSuccess(response);
     });
   }
+
+  /**
+   * Add "__amp_source_origin" query parameter to the URL. Ideally, we'd be
+   * able to set a header (e.g. AMP-Source-Origin), but this will force
+   * preflight request on all CORS request.
+   * @param {!Window} win
+   * @param {string} url
+   * @return {string}
+   */
+  getCorsUrl(win, url) {
+    const sourceOrigin = getSourceOrigin(win.location.href);
+    const parsedUrl = parseUrl(url);
+    const query = parseQueryString(parsedUrl.search);
+    user.assert(!(SOURCE_ORIGIN_PARAM in query),
+        'Source origin is not allowed in %s', url);
+    return addParamToUrl(url, SOURCE_ORIGIN_PARAM, sourceOrigin);
+  }
+
+  /**
+   * @param {!ArrayBuffer} bytes
+   * @return {!Promise<string>}
+   */
+  utf8FromArrayBuffer(bytes) {
+    if (window.TextDecoder) {
+      return Promise.resolve(new TextDecoder('utf-8').decode(bytes));
+    }
+    return new Promise(function(resolve, unusedReject) {
```

Comment:
Is there an error event we can listen to? Otherwise, we can remove the `unusedReject` param.